Tests/test-dataframe.lisp

```diff
@@ -0,0 +1,554 @@
+        (5 (let* ((length (funcall (gen-integer :min 1 :max (length rows))))
+                  (wrong (random length))
+                  (array (make-array length)))
+             (loop for i from 1 below length
+                   do (setf (elt array i)
+                            (funcall gen-index))
+                   finally (setf (elt array wrong)
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+(defun gen-invalid-col-slice (data-frame)
+  (let ((gen-index (gen-valid-col-index data-frame))
+        (gen-index* (gen-invalid-col-index data-frame))
+        (cols (cols data-frame)))
+    (lambda ()
+      (case (random 6)
+        (0 nil)
+        (1 (cons t (funcall gen-index*)))
+        (2 (cons (funcall gen-index*) t))
+        (3 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index)))
+             (cond ((and (integerp i1)
+                         (integerp i2))
+                    (if (< i1 i2)
+                        (cons i2 i1)
+                        :invalid))
+                   ((integerp i1)
+                    (if (< i1 (position i2 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid))
+                   ((integerp i2)
+                    (if (> i2 (position i1 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid))
+                   (t
+                    (if (< (position i1 cols :test #'string=)
+                           (position i2 cols :test #'string=))
+                        (cons i2 i1)
+                        :invalid)))))
+        (4 (let ((i1 (funcall gen-index))
+                 (i2 (funcall gen-index*))
+                 (i3 (funcall gen-index*)))
+             (case (random 3)
+               (0 (cons i1 i2))
+               (1 (cons i2 i1))
+               (2 (cons i2 i3)))))
+        (5 (let* ((length (funcall (gen-integer :min 1 :max (length cols))))
+                  (wrong (random length))
+                  (array (make-array length)))
+             (loop for i from 0 below length
+                   do (setf (elt array i)
+                            (funcall gen-index))
+                   finally (setf (elt array wrong)
+                                 (funcall gen-index*))
+                           (return array))))))))
+
+
+
+(defun gen-valid-function (who)
+  (lambda ()
+    (case (random 2)
+      ;; constant
+      (0 (constantly 42))
+      ;; combination of inputs
+      (1 (case who
+           (row (lambda (row-index row)
+                  (declare (ignore row-index))
+                  (if (> (length row) 2)
+                      (cons (elt row 0) (elt row 1))
+                      42)))
+           (col (lambda (col-index col-name value)
+                  (declare (ignore col-index))
+                  (cons col-name value)))
+           (map (lambda (row-index row col-index col-name value)
```

Comment:
is row/col/map always used for mapping? I think that in that case they should be named "map-row", "map-col" and "map-all" -- IMO it will be easier to grok tests for new readers.

Reply:
The who=row is also used for add-cols! test (indirectly through gen-valiid-name-fun-pairs). I originally had a separate gen-(in)valid-function-X for each function (a total of 8), but the code was so similar, it looked ugly. I will work on the naming and some comments. Do you think it would actually be better with separate functions? Maybe some kind of macro for the common code and then separate to deal with each function (e.g. add-cols!, map-df-X)?

Reply:
I think that these two functions are perfectly fine (having 8 of them would be clearly worse), just add a comment which describes what's going on and rename valid cases.

Reply:
See new commit. Changed to use name of function being tested.